[DASC-27] Major revision for upon the reviews received

furkanmtorun opened the pull request:
---

**Summary:** Major revision visiting the comments from the reviewers (towards to OmicLearn v1.1)
**JIRA Ticket:** https://omicera.atlassian.net/browse/DASC-27 

---

## Changes:

- [X] Upgraded the Streamlit version to be able to use new features.
- [X] Upgraded the other dependencies
- [X] Tried to simplify the code and re-factor some parts
- [X] Worked on improvement on UI and UX side
- [X] Implement a section where users can perform PCA and hierarchical clustering
- [X] Create the Wiki page and add a section for PCA and hierarchical clustering, update links
- [X] On the UI side, put disclaimers that the service is provided for data exploration purposes and not for use in the production.
- [X] With the new version of Streamlit, there is an issue with the `"Session History"` section I previously implemented. This is also resolved. 

## Note the followings:
- As for linting and coding style, planning to submit another PR to improve the code quality. You might save your comments on this for the next PR.

Best,
Furkan

straussmaximilian commented on omic_learn.py:
```diff
@@ -1,20 +1,28 @@
+APP_TITLE = "OmicLearn — ML platform for biomarkers"
+st.set_page_config(
+    page_title = APP_TITLE, 
+    page_icon = Image.open('./utils/omic_learn_black.png'), 
```
Cool idea to change the favicon. We could also create a dedicated smaller icon for this.

furkanmtorun replied:
Great point! Thanks a lot! So, now, used 48x48 ico file as favicon!

straussmaximilian commented on omic_learn.py:
```diff
@@ -180,88 +114,89 @@ def checkpoint_for_data_upload(state, record_widgets):
-        state['class_1'] = multiselect("Select Class 1:",
-                                       [_ for _ in unique_elements_lst if _ not in state.class_0], default=None)
-        state['remainder'] = [_ for _ in state.not_proteins if _ is not state.target_column]
+        with st.beta_expander("Classification target (*Required)"):
```
One could make the *Required in bold (***Required**), like you did with the **Warning**

furkanmtorun replied:
Actually, I have tried this one as it would be really good for both UI & UX. However, neither way is working since it interprets them as normal characters :( 

![image](https://user-images.githubusercontent.com/49681382/120076738-ce6d2000-c0af-11eb-8ffb-b9259b751cff.png)

straussmaximilian commented on requirements.txt:
```diff
@@ -1,8 +1,9 @@
+plotly==4.14.3
+kaleido==0.2.1
+XlsxWriter==1.4.0
+seaborn==0.11.1
```
it seems that we are not using Seaborn?

furkanmtorun replied:
Yes, forgot to remove it. Done.

straussmaximilian reviewed (changes requested):
Hi, 

I think you did a fantastic job with the new column expander layout. This is pretty cool, and I think you solved the problem on how to highlight what is required and what not excellently. Excellent job!

Some minor comments:

- [ ] When clicking the link to the license agreement, one will be directed to the repo main page and not the document
- [x]  When you have uploaded / selected your data and nothing else, there will be a message to start with defining classes. However, the first **required** row is to select the classification target, so maybe we can change this to start with selecting a classification target.

Some additional suggestions:

- The beta-expander is providing already some conditionality, so one could think of skipping the additional checkboxes (Yes, I would like to exclude features / manually select features)
- I would move the EDA part from the sidebar to the main window after the create subset and before classification target - with the option to get a plot before even running the analysis. The idea would be that one could get much quicker feedback about the data before setting everything and for example, create a data subset.
- Streamlit has now the option for themes. Thise could be done via setting the config.toml, however another way to do this is to provide arguments when launching, e.g. `streamlit run file.py --theme.primaryColor=#18212b --theme.font='sans serif'" --theme.textColor=#262730" --theme.secondaryBackgroundColor=#f0f2f6" --theme.backgroundColor=#FFFFFF"`

Maybe this could help to set the button style.

Bugs I found:
I also encountered the session-state bug that you mentioned
<img width="1280" alt="Screenshot 2021-05-29 at 12 07 58" src="https://user-images.githubusercontent.com/16521288/120066890-073dd280-c079-11eb-96d6-b3a440fab2bb.png">

For a test, the PCA/ Hierarchical clustering was not working for me.
This is what I tried:
- Start
- Select Alzheimer
- Define _clinical AD diagnosis
- Define classes -> non-AD / AD
- Select PCA with 2
- Run analysis

<img width="1280" alt="Screenshot 2021-05-29 at 12 05 26" src="https://user-images.githubusercontent.com/16521288/120066893-0ad15980-c079-11eb-8fae-7f953f57ac40.png">

furkanmtorun commented:
Dear Max (@straussmaximilian ),

I pushed the new commits after your comments.
So, I have addressed the following points:

- Update license agreement and link it to the UI
- Update favicon with a smaller image
- Remove unnecessary checkboxes
- Data selection warnings are updated
- Improve UI & UX 
- Relocate and refactor EDA part
- Fix PCA section (With Alzheimer dataset, it is freezing (most prob.) due to capacity of the instance currently used)
- Refactor Hierar. Clustering section using dash_bio as it makes everything easier

Other notes:
- Prefer to use CSS as it has more control on the page over the theme settings
- Will work on the `"session history" section in the next PR.

Hope those updates find you well

Best,
Furkan

straussmaximilian commented:
Hi Furkan,

great PR! Thinks are going good in the right direction!

I have a couple more small suggestions:

### Suggestions for text change:

`By uploading a file, you agree that you accepting the license agreement.`

-> `By uploading a file, you agree that you are accepting the license agreement. Note: We do not save the data you upload via the FileUploader; it is only stored temporarily in RAM to perform the calculations.`

After the selection of a class, there is a warning. Technically we already started, so maybe one can change the text accordingly:

`WARNING: Start with defining classes.` -> `WARNING: Continue with defining classes.`

On this note: The Warning for `WARNING: Start with defining classes.` Is red while as the missing value warning: `WARNING: Found 69255 missing values. Use missing value imputation or xgboost classifier.` is yellow.

I just checked and streamlit offers the following styling elements:
<img width="751" alt="Screenshot 2021-06-07 at 10 20 49" src="https://user-images.githubusercontent.com/16521288/120983641-2ecc2380-c77a-11eb-9469-c826687f0e8e.png">

We could adjust them to look similar (e.g. consider the missing value warning and start with defining classes all as info and make them blue) - what do you think? Would be just for extra points of course.
 
### EDA
I selected the Alzheimer dataset, selected the classes, clicked PCA and pressed Perform EDA. The upper right corner said running. I waited for 5 minutes but nothing happened. One could also use st.spinner to indicate something is happening

The hierarchical clustering did work but slowed down the application. When clicking `Run analysis` afterwards things were a bit slow. Maybe we could add functionality to `Run analysis` to remove existing EDA plots for speed up? Currently they dissapear when pressing `Run analysis` anyhow.

For me I had problems visualising the hierarchical clustering plot: 
<img width="1357" alt="Screenshot 2021-06-06 at 23 30 56" src="https://user-images.githubusercontent.com/16521288/120984291-cfbade80-c77a-11eb-9e61-9dbc55a28d60.png">
<img width="1357" alt="Screenshot 2021-06-06 at 23 31 12" src="https://user-images.githubusercontent.com/16521288/120984296-d0ec0b80-c77a-11eb-8542-208d1599800c.png">


### Performance
Overall performance: Somehow, I have the feeling that things are much slower than before. At first, I got a suggestion from streamlit to install watchdog, which I did. This made things a bit faster, so I think we should add this to the requirements. One oddity was that when starting OmicLearn for the first time it was slow and the more I used it, the faster it got. Maybe this was just a temporary issue - did you also experience this? Otherwise, we could add a note that OmicLearn might be slow when starting for the first time.

Other than that well done! Looking forward to the session history bug.

furkanmtorun commented:
Hi Max,

Thanks for your comment. The points you addressed are also awesome!

Here is what these new commits bring:

- [X] I updated the text for the "data uploading section".
- [X] I totally agree with the use of the correct element for notifications (st.error, st.info, st.warning).  

  In this respect, I use the following strategy:
  
  **Info**: This only serves for letting know the user about the thing. I used for general info texts (app details at the top, citation). Additonaly, I classified "missing value" part as info since the user still continue working by selecting "xgboost". 
  
  **Warning**: This serves to warn the user about either lack of selection or misconfiguration (i.e. If user did not select the classification target, it **warns** the user.) Here, user still can continue working by filling the required areas or update the misconfigured places.
  
  **Error**: This st.error element only will be shown if the selection or configuration make the app terminated or failed.

**Will check the following things, update or leave new comment:**:

- [ ] As for PCA, as I told you, I tried with random other datasets, it was working. I assumed that it might be related to instance capacity for Alzheimer's dataset. I will check what is the case with it.

- [ ] As for hierarchical clustering, I actually made the ticks label shorter beforehand. However, it did not solve the problem. Also, for heatmap side -coloring-. seems buggy. Will check this too.   

- [ ] In terms of performance, exactly. I observed the same issues when I upgraded the streamlit version from 0.63 to 0.81.1, even without doing any extra things. I will test this "watchdog | https://pypi.org/project/watchdog/" package. 

Best,
Furkan

SecretLake commented on utils/helper.py:
```diff
@@ -44,7 +46,84 @@ def wrapper(label, *args, **kwargs):
+    widget_values = objdict()
+    record_widgets = objdict()
+
+    # Sidebar widgets
```
Just an idea, this could be re-written to be more concise:

elements = {
'button_': st.sidebar.button,
'slider_': st.sidebar.slider,
etc.
}

for key, value in elements.items():
record_widgets[key] = make_recording_widget(value, widget_values)

furkanmtorun replied:
Hey Artem,

Thanks for addressing this point! Just refactored that part now!

SecretLake replied:
Great stuff!

SecretLake reviewed (approved):
Hi Furkan, looks good to me!

straussmaximilian commented:
Hi Furkan,

Good progress. I just tested OmicLearn and found the following things:

---
- Add watchdog to requirements
---
Suggested text changes:

`Classification target refers to the variable that the classifier should be able to distinguish. Then, the 2 classes from the values in this column should be defined.`

->

`Classification target refers to the column that contains the variables that are used two distinguish the two classes. In the next section, the unique values of this column can be used to define the two classes. `

---
`As for a binary classification task, two options in _clinical AD diagnosis column should be defined for the outcome of the classifier. By assigning multiple values to a class, multiple combinations of classifications can be tested.`

-> {} is placeholder for the variable.

`For a binary classification task, one needs to define two classes based on the unique values in the {_clinical AD diagnosis} task column. It is possible to assign multiple values for each class.`

---
`Exploratory data analysis is performed on the whole dataset for providing more insight. For more information, please visit the dedicated Wiki page.`

->

`Exploratory data analysis is performed on the whole dataset to provide more insight. For more information, please visit the dedicated Wiki page.`

---
Clicking EDA now works for me, however I think we should adjust the PCA plot a bit.
When hovering I see the class (e.g. _clinical AD diganosis=AD). Here it would be nice to get. e.g. the sample number to trace back which file it was.

There seems to be something off with the hierarchical clustering:
- the range goes from -4 to 14 - I think this should be [-1, 1]
- there doesn't seem to be a self-correlation of the samples 

See attached a sample correlation map from the COVID study:

![Screenshot 2021-06-28 at 13 52 12](https://user-images.githubusercontent.com/16521288/123632292-0fc02f00-d818-11eb-8e79-6198f1c905f0.png)

---
![Screenshot 2021-06-28 at 13 31 15](https://user-images.githubusercontent.com/16521288/123632508-4dbd5300-d818-11eb-897b-7ffa7388e51e.png)


The Top features from the classifier seems to be messed up (x-axis seems to be categorical)
Note the Remainder is 0.354 so it should be the largest bar but it is actually the smallest.

---

Suggested text change:

`Info: Mean precision and Mean recall values provided in the table above are calculated from each individual splits shown in the confusion matrix, not "Sum of all splits" matrix.`

-> 

`Info: Mean precision and Mean recall values provided in the table above are calculated as the mean of all individual splits shown in the confusion matrix, not the "Sum of all splits" matrix.`


Best,


Max

furkanmtorun commented:
Hi Max,

- I have updated the texts you pointed out.
- Also, fixed the issue with feature importance chart. The reason behind the problem is not related to "categorical" axis, rather it is about the data type.So, convertion all of them to float32 fixed it.
- I got rid of dash_bio.Clustergram as it is not working properly. So, switched back to original native Plotly chart after checking the examples.

Hope those update find you well.

Cheers,
Furkan
